File: 0312-burst-balloons/0312-burst-balloons.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums.insert(0, 1)
        nums.append(1)
        
        memo = {}
        def dfs(left, right):
            if left > right:
                return 0
            
            if (left, right) in memo:
                return memo[(left, right)]
            ans = 0
            for k in range(left, right + 1):
                ans = max(ans, dfs(left, k - 1) + (nums[left - 1] * nums[k] * nums[right + 1]) + dfs(k + 1, right))
            
            memo[(left, right)] = ans
            return ans
        
        return dfs(1, len(nums) - 2)

File: 0312-burst-balloons/0312-burst-balloons.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        nums.insert(nums.begin(), 1);
        nums.push_back(1);
        vector<vector<int>> dp(305, vector<int>(305, -1));
        
        function<int(int, int)> dfs;
        dfs = [&](int left, int right) {
            if (left > right)
                return 0;
            if (dp[left][right] != -1)
                return dp[left][right];
            int ans = 0;
            for (int k = left; k <= right; k++)
                ans = max(ans, dfs(left, k - 1) + (nums[left - 1] * nums[k] * nums[right + 1]) + dfs(k + 1, right));
            
            return dp[left][right] = ans;
        };
        
        return dfs(1, nums.size() - 2);
    }
};